common:
    lidar_topic:  "/os1_cloud_node1/points"
    imu_topic:  "/os1_cloud_node1/imu"
    image_topic:  "/left/image_raw"
    image_type:  RGB8    # 1 RGB8   2 COMPRESSED
    point_filter_num: 4
    time_sync_en: false         # ONLY turn on when external time synchronization is really not possible
    gravity_acc: [ 0.0, 0.0, 9.81]
    
lidar_parameter:
    lidar_type: 3                # 1 for Livox serials LiDAR, 2 for Velodyne LiDAR, 3 for ouster LiDAR, 
    N_SCANS: 16
    SCAN_RATE: 20                # only need to be set for velodyne, unit: Hz,
    time_unit: 3            # the unit of time/t field in the PointCloud2 rostopic: 0-second, 1-milisecond, 2-microsecond, 3-nanosecond.
    blind: 4
    fov_degree:    180
    det_range:     100.0

imu_parameter:
    acc_cov: 0.1
    gyr_cov: 0.1
    b_acc_cov: 0.0001
    b_gyr_cov: 0.0001
    time_diff_enable: false
    
camera_parameter:
    image_width: 752
    image_height: 480
    camera_intrinsic: [ 425.0259, 0.0, 386.0152,
                        0.0, 426.7976, 241.9130,
                        0.0, 0.0, 1.0 ]
    camera_dist_coeffs: [ -0.2881, 0.0746, 7.7845e-04, -2.2779e-04, 0.0 ]  #k1, k2, p1, p2, k3

extrinsic_parameter:
    extrinsic_enable:  false      # true: enable the online estimation of IMU-LiDAR extrinsic,
    extrinsic_t_imu_lidar: [ 0.0, 0.0, 0.0 ]
    extrinsic_R_imu_lidar: [ 1, 0, 0, 
                             0, 1, 0, 
                             0, 0, 1 ]
    extrinsic_t_imu_camera: [ 0.0555294, -0.124313, -0.0388531 ]
    extrinsic_R_imu_camera: [ 0.0218308, -0.0131205, 0.999675,
                              0.999759, 0.00230088, -0.0218024,
                             -0.00201407, 0.999912, 0.0131676 ]

odometry_options:
    voxel_size: 0.1 # The voxel size for the grid sampling of the new frame (before keypoints extraction)
    sample_voxel_size: 1.5 # The size of a voxel for the selection of `keypoints` by grid sampling
    max_distance: 2000.0 # The threshold of the distance to suppress voxels from the map
    max_num_points_in_voxel: 20 # The maximum number of points per voxel of the map
    init_num_frames: 20
    min_distance_points: 0.15
    distance_error_threshold: 100.0 # The motion of the sensor between two frames which is considered erroneous (stops the odometry)
    motion_compensation: CONSTANT_VELOCITY # The profile of the motion compensation (IMU, CONSTANT_VELOCITY)
    initialization: INIT_IMU # [INIT_IMU, INIT_CONSTANT_VELOCITY]

icp_options:
    size_voxel_map: 1.0 # The voxel size of in the voxel map
    num_iters_icp: 5 # The number of iterations of the ICP
    min_number_neighbors: 20 # The minimum number of neighbor points to define a valid neighborhood
    voxel_neighborhood: 1
    max_number_neighbors: 20
    max_dist_to_plane_ct_icp: 0.3
    threshold_orientation_norm: 0.1 # Threshold on orientation changes (in degrees) for early termination of the ICP
    threshold_translation_norm: 0.01 # Threshold on distance changes (in m) for early termination of the ICP
    debug_print: false
    num_closest_neighbors: 1
    min_num_residuals: 200 # The minimum number of residuals for a valid ICP problem
    max_num_residuals: 600 # The maximum number of residuals considered (if more keypoints exist, residuals are randomly sampled)

map_options:
    size_voxel_map: 0.1 # 0.01 The voxel size of in the voxel map for visualization
    max_num_points_in_voxel: 50 # 1 The maximum number of points per voxel of the map for visualization
    min_distance_points: 0.01 # The minimum distance of two points for visualization
    add_point_step: 1  # The point step of adding point to color map. [default = 4]
    pub_point_minimum_views: 1  # Publish points which have been render up to "pub_pt_minimum_views" time. [default = 3]